{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "id": "voq_EIyaOPzX",
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# make sure TF is below 2.16!\n",
    "# Otherwise the model won't be able to load.\n",
    "import tensorflow as tf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.15.0\n"
     ]
    }
   ],
   "source": [
    "print(tf.__version__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "id": "vkkgAZHvWhh8",
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mon Dec  2 15:23:19 2024       \n",
      "+---------------------------------------------------------------------------------------+\n",
      "| NVIDIA-SMI 535.183.01             Driver Version: 535.183.01   CUDA Version: 12.2     |\n",
      "|-----------------------------------------+----------------------+----------------------+\n",
      "| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |\n",
      "| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |\n",
      "|                                         |                      |               MIG M. |\n",
      "|=========================================+======================+======================|\n",
      "|   0  NVIDIA GeForce RTX 2080 Ti     Off | 00000000:88:00.0 Off |                  N/A |\n",
      "| 30%   25C    P8              19W / 250W |      1MiB / 11264MiB |      0%      Default |\n",
      "|                                         |                      |                  N/A |\n",
      "+-----------------------------------------+----------------------+----------------------+\n",
      "|   1  NVIDIA GeForce RTX 2080 Ti     Off | 00000000:89:00.0 Off |                  N/A |\n",
      "| 30%   25C    P8               7W / 250W |      1MiB / 11264MiB |      0%      Default |\n",
      "|                                         |                      |                  N/A |\n",
      "+-----------------------------------------+----------------------+----------------------+\n",
      "|   2  NVIDIA GeForce RTX 2080 Ti     Off | 00000000:8A:00.0 Off |                  N/A |\n",
      "| 30%   26C    P8              21W / 250W |      1MiB / 11264MiB |      0%      Default |\n",
      "|                                         |                      |                  N/A |\n",
      "+-----------------------------------------+----------------------+----------------------+\n",
      "|   3  NVIDIA GeForce RTX 2080 Ti     Off | 00000000:8B:00.0 Off |                  N/A |\n",
      "| 30%   26C    P8               8W / 250W |      1MiB / 11264MiB |      0%      Default |\n",
      "|                                         |                      |                  N/A |\n",
      "+-----------------------------------------+----------------------+----------------------+\n",
      "|   4  NVIDIA GeForce RTX 2080 Ti     Off | 00000000:DB:00.0 Off |                  N/A |\n",
      "| 30%   25C    P8               7W / 250W |      1MiB / 11264MiB |      0%      Default |\n",
      "|                                         |                      |                  N/A |\n",
      "+-----------------------------------------+----------------------+----------------------+\n",
      "|   5  NVIDIA GeForce RTX 2080 Ti     Off | 00000000:DC:00.0 Off |                  N/A |\n",
      "| 30%   26C    P8               9W / 250W |      1MiB / 11264MiB |      0%      Default |\n",
      "|                                         |                      |                  N/A |\n",
      "+-----------------------------------------+----------------------+----------------------+\n",
      "|   6  NVIDIA GeForce RTX 2080 Ti     Off | 00000000:DD:00.0 Off |                  N/A |\n",
      "| 30%   26C    P8              21W / 250W |      1MiB / 11264MiB |      0%      Default |\n",
      "|                                         |                      |                  N/A |\n",
      "+-----------------------------------------+----------------------+----------------------+\n",
      "|   7  NVIDIA GeForce RTX 2080 Ti     Off | 00000000:DE:00.0 Off |                  N/A |\n",
      "| 30%   25C    P8              16W / 250W |      1MiB / 11264MiB |      0%      Default |\n",
      "|                                         |                      |                  N/A |\n",
      "+-----------------------------------------+----------------------+----------------------+\n",
      "                                                                                         \n",
      "+---------------------------------------------------------------------------------------+\n",
      "| Processes:                                                                            |\n",
      "|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |\n",
      "|        ID   ID                                                             Usage      |\n",
      "|=======================================================================================|\n",
      "|  No running processes found                                                           |\n",
      "+---------------------------------------------------------------------------------------+\n"
     ]
    }
   ],
   "source": [
    "# GPU info:\n",
    "!nvidia-smi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "id": "4mHxe2YXWhh9"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Num GPUs Available:  1\n"
     ]
    }
   ],
   "source": [
    "# Assign GPU to use:\n",
    "GPU_id = '7'\n",
    "import os\n",
    "os.environ[\"CUDA_VISIBLE_DEVICES\"]=GPU_id\n",
    "# check:\n",
    "import tensorflow as tf\n",
    "print(\"Num GPUs Available: \", len(tf.config.experimental.list_physical_devices('GPU')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "id": "98NU5NCGWhh9"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GPU working\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-12-02 15:29:57.642846: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1929] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 9803 MB memory:  -> device: 0, name: NVIDIA GeForce RTX 2080 Ti, pci bus id: 0000:de:00.0, compute capability: 7.5\n"
     ]
    }
   ],
   "source": [
    "# check:\n",
    "sess = tf.compat.v1.Session()\n",
    "gpu_devices = tf.config.list_physical_devices('GPU')\n",
    "if len(gpu_devices) > 0:\n",
    "    print(\"GPU working\")\n",
    "else:\n",
    "    print(\"GPU not working\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "id": "yjElLEtyWhh-"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "device name: /physical_device:GPU:0\n",
      "device type: GPU\n"
     ]
    }
   ],
   "source": [
    "# check:\n",
    "for device in gpu_devices:\n",
    "    print(\"device name:\", device.name)\n",
    "    print(\"device type:\", device.device_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "Xkr_7oyPWhh-"
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "L5_8mer_df = pd.read_table('/content/L5_2-8mer.tsv')\n",
    "L5_8mer_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "id": "xcC6RT4Ca0n1",
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: logomaker in ./miniconda3/lib/python3.11/site-packages (0.8)\n",
      "Requirement already satisfied: numpy in ./miniconda3/lib/python3.11/site-packages (from logomaker) (1.26.4)\n",
      "Requirement already satisfied: matplotlib in ./miniconda3/lib/python3.11/site-packages (from logomaker) (3.9.2)\n",
      "Requirement already satisfied: pandas in ./miniconda3/lib/python3.11/site-packages (from logomaker) (2.2.2)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in ./miniconda3/lib/python3.11/site-packages (from matplotlib->logomaker) (1.2.1)\n",
      "Requirement already satisfied: cycler>=0.10 in ./miniconda3/lib/python3.11/site-packages (from matplotlib->logomaker) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in ./miniconda3/lib/python3.11/site-packages (from matplotlib->logomaker) (4.53.1)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in ./miniconda3/lib/python3.11/site-packages (from matplotlib->logomaker) (1.4.5)\n",
      "Requirement already satisfied: packaging>=20.0 in ./miniconda3/lib/python3.11/site-packages (from matplotlib->logomaker) (24.1)\n",
      "Requirement already satisfied: pillow>=8 in ./miniconda3/lib/python3.11/site-packages (from matplotlib->logomaker) (10.4.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in ./miniconda3/lib/python3.11/site-packages (from matplotlib->logomaker) (3.1.2)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in ./miniconda3/lib/python3.11/site-packages (from matplotlib->logomaker) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in ./miniconda3/lib/python3.11/site-packages (from pandas->logomaker) (2024.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in ./miniconda3/lib/python3.11/site-packages (from pandas->logomaker) (2024.1)\n",
      "Requirement already satisfied: six>=1.5 in ./miniconda3/lib/python3.11/site-packages (from python-dateutil>=2.7->matplotlib->logomaker) (1.16.0)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install logomaker"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "id": "RAQNScryWhh_"
   },
   "outputs": [],
   "source": [
    "# 8-mer mutagenesis:\n",
    "\n",
    "# Function to generate a list of all possible mutations for a seq:\n",
    "def all_possible_mutations(dna_seq):\n",
    "    mutated_seqs = []\n",
    "    for i in range(len(dna_seq)):\n",
    "        for nucleotide in [\"A\", \"T\", \"C\", \"G\"]:\n",
    "            if nucleotide != dna_seq[i]:\n",
    "                mutated_seq = dna_seq[:i] + nucleotide + dna_seq[i+1:]\n",
    "                mutated_seqs.append(mutated_seq)\n",
    "    #\n",
    "    return mutated_seqs\n",
    "\n",
    "\n",
    "# functions to get a df of the regulatory relevance of each nucleotide of 8-mer:\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "def kmer_saliency_df(kmer = 'GGTAAGTA', target = 'delta.log2expression'):\n",
    "    # generate all point mutations:\n",
    "    all_mutants = all_possible_mutations(kmer)\n",
    "    # add the original seq to the mutant list:\n",
    "    all_mutants.insert(0, kmer)\n",
    "    # create empty df:\n",
    "    kmer_values = pd.DataFrame(columns=['kmer', 'value'])\n",
    "    # find values by each k-mer:\n",
    "    for the_kmer in all_mutants:\n",
    "        match = L5_8mer_df.loc[L5_8mer_df['kmer'] == the_kmer, target]\n",
    "        new_row = pd.DataFrame({'kmer': [the_kmer], 'value': [match.values[0]]})\n",
    "        kmer_values = pd.concat([kmer_values, new_row], ignore_index=True)\n",
    "    # calculate delta:\n",
    "    values = kmer_values['value'].values\n",
    "    deltas = values[0] - values[1:]\n",
    "    # median deltas of each original nucleotide/position:\n",
    "    delta_medians = []\n",
    "    for i in range(3, len(deltas)+1, 3):\n",
    "        median = np.median(deltas[i-3:i])\n",
    "        delta_medians.append(median)\n",
    "    # form the final data.frame suitable for logomaker:\n",
    "    position_list = list(kmer)\n",
    "    final_df = pd.DataFrame(columns=['A', 'C', 'G', 'T'])\n",
    "    for i, letter in enumerate(position_list):\n",
    "        final_df.at[i, letter] = delta_medians[i]\n",
    "    final_df = final_df.fillna(0)\n",
    "    final_df = final_df.astype('float64')\n",
    "    # change row index to 1,2,3...:\n",
    "    final_df.index = range(1, len(final_df) + 1)\n",
    "    #\n",
    "    return final_df\n",
    "\n",
    "\n",
    "# function to plot nucleotide saliencies:\n",
    "import logomaker\n",
    "import matplotlib.pyplot as plt\n",
    "def plot_saliency(df, negative=False,\n",
    "                  start=None, end=None, figsize=[8,2],\n",
    "                  xticks=False, yticks=False,\n",
    "                  spines=False, ylim=None):\n",
    "    # make Figure and Axes objects:\n",
    "    fig, ax = plt.subplots(1,1,figsize=figsize)\n",
    "    # limit x range, if defined:\n",
    "    if start is not None and end is not None:\n",
    "        df = df[start : end+1]\n",
    "    elif start is not None:\n",
    "        df = df[start : ]\n",
    "    elif end is not None:\n",
    "        df = df[ : end+1]\n",
    "    # flip saliencies if defined:\n",
    "    if negative == True:\n",
    "        df = -df\n",
    "    #\n",
    "    logo = logomaker.Logo(df, ax=ax)\n",
    "    #\n",
    "    if ylim is not None:\n",
    "        logo.ax.set_ylim(ylim)\n",
    "    #\n",
    "    if spines==False:\n",
    "        logo.style_spines(visible=False)\n",
    "    #\n",
    "    if xticks==False:\n",
    "        ax.set_xticks([])\n",
    "    #\n",
    "    if yticks==False:\n",
    "        ax.set_yticks([])\n",
    "    #\n",
    "    return logo.fig.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "kum-RaV8Whh_"
   },
   "outputs": [],
   "source": [
    "# k-mer motif:\n",
    "kmer = 'AGGTAAGT'\n",
    "target = 'delta.log2expression' # delta.log2expression / delta.log2export\n",
    "\n",
    "the_kmer_saliency_df = kmer_saliency_df(kmer=kmer, target=target)\n",
    "the_kmer_saliency_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "AyfKmLi3WhiA"
   },
   "outputs": [],
   "source": [
    "plot_saliency(df=the_kmer_saliency_df, negative=True, figsize=[3,2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: seaborn in ./miniconda3/lib/python3.11/site-packages (0.13.2)\n",
      "Requirement already satisfied: numpy!=1.24.0,>=1.20 in ./miniconda3/lib/python3.11/site-packages (from seaborn) (1.26.4)\n",
      "Requirement already satisfied: pandas>=1.2 in ./miniconda3/lib/python3.11/site-packages (from seaborn) (2.2.2)\n",
      "Requirement already satisfied: matplotlib!=3.6.1,>=3.4 in ./miniconda3/lib/python3.11/site-packages (from seaborn) (3.9.2)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in ./miniconda3/lib/python3.11/site-packages (from matplotlib!=3.6.1,>=3.4->seaborn) (1.2.1)\n",
      "Requirement already satisfied: cycler>=0.10 in ./miniconda3/lib/python3.11/site-packages (from matplotlib!=3.6.1,>=3.4->seaborn) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in ./miniconda3/lib/python3.11/site-packages (from matplotlib!=3.6.1,>=3.4->seaborn) (4.53.1)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in ./miniconda3/lib/python3.11/site-packages (from matplotlib!=3.6.1,>=3.4->seaborn) (1.4.5)\n",
      "Requirement already satisfied: packaging>=20.0 in ./miniconda3/lib/python3.11/site-packages (from matplotlib!=3.6.1,>=3.4->seaborn) (24.1)\n",
      "Requirement already satisfied: pillow>=8 in ./miniconda3/lib/python3.11/site-packages (from matplotlib!=3.6.1,>=3.4->seaborn) (10.4.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in ./miniconda3/lib/python3.11/site-packages (from matplotlib!=3.6.1,>=3.4->seaborn) (3.1.2)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in ./miniconda3/lib/python3.11/site-packages (from matplotlib!=3.6.1,>=3.4->seaborn) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in ./miniconda3/lib/python3.11/site-packages (from pandas>=1.2->seaborn) (2024.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in ./miniconda3/lib/python3.11/site-packages (from pandas>=1.2->seaborn) (2024.1)\n",
      "Requirement already satisfied: six>=1.5 in ./miniconda3/lib/python3.11/site-packages (from python-dateutil>=2.7->matplotlib!=3.6.1,>=3.4->seaborn) (1.16.0)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install seaborn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "id": "1BIznd3tWhiA"
   },
   "outputs": [],
   "source": [
    "# k-mer context dependencies:\n",
    "\n",
    "\n",
    "# Define the model:\n",
    "model = '/rd4/users/liangn/L5-220528_em5-LSTM64x32x0.5-64x0.5-rep4.hdf5'\n",
    "model_x_length = 46\n",
    "optimal_x_length = 45\n",
    "\n",
    "\n",
    "# Function to convert a DNA sequence to vector:\n",
    "vocab = ['pad','N','A','T','C','G']\n",
    "char2idx = {u:i for i, u in enumerate(vocab)}\n",
    "def vectorize_dna_seq(dna_seq):\n",
    "    vectorized_dna_seq = [char2idx[char] for char in dna_seq]\n",
    "    return vectorized_dna_seq\n",
    "\n",
    "\n",
    "# Function to convert a list of DNA into x array for ANN inputs:\n",
    "from tensorflow.keras.preprocessing.sequence import pad_sequences\n",
    "def prepare_x(dna_list, x_lenth):\n",
    "    x = list(map(vectorize_dna_seq, dna_list))\n",
    "    x = pad_sequences(x, maxlen=x_lenth, padding='post')\n",
    "    #\n",
    "    return x\n",
    "\n",
    "\n",
    "# function to split a string into k-mers:\n",
    "def kmerize(string, k):\n",
    "    return [string[i:i+k] for i in range(len(string)-k+1)]\n",
    "\n",
    "\n",
    "# function to predict big seq with sliding windows:\n",
    "import pandas as pd\n",
    "from keras.models import load_model\n",
    "import numpy as np\n",
    "def predict_big(big_seqs, k=optimal_x_length, model_x_length=model_x_length, model=model):\n",
    "    # form data.frame:\n",
    "    # if big_seqs only had one, convert to a list as well:\n",
    "    if isinstance(big_seqs, str):\n",
    "        big_seqs = [big_seqs]\n",
    "    seq_df = pd.DataFrame({'seq': big_seqs})\n",
    "    # split each seq into k-mers, mark indexes:\n",
    "    seq_df = seq_df['seq'].apply(lambda x: kmerize(x, k))\n",
    "    seq_df = seq_df.apply(pd.Series)\n",
    "    seq_df = seq_df.stack().reset_index(level=1, drop=True).to_frame('seq')\n",
    "    # prepare x for predictions:\n",
    "    x = seq_df['seq'].apply(vectorize_dna_seq)\n",
    "    x = pad_sequences(x, maxlen=model_x_length, padding='post')\n",
    "    # predict:\n",
    "    Model = load_model(model)\n",
    "    y_pred = Model.predict(x)\n",
    "    # take group means of the same indexes:\n",
    "    unique_index = np.unique(seq_df.index)\n",
    "    y_pred_mean = np.zeros((len(unique_index), y_pred.shape[1]))\n",
    "    for i in range(len(unique_index)):\n",
    "        y_pred_mean[i] = np.mean(y_pred[seq_df.index == unique_index[i]], axis=0)\n",
    "    #\n",
    "    return y_pred_mean\n",
    "\n",
    "\n",
    "# function to get a df of the regulatory relevance of each nucleotide of a seq:\n",
    "from keras.models import load_model\n",
    "import pandas as pd\n",
    "def saliency_df(seq, optimal_x_length=optimal_x_length, model_x_length=model_x_length, model=model,\n",
    "                target='expression'):\n",
    "    # generate all point mutations:\n",
    "    all_mutants = all_possible_mutations(seq)\n",
    "    # add the original seq to the mutant list:\n",
    "    all_mutants.insert(0, seq)\n",
    "    # check seq length:\n",
    "    seq_len = len(seq)\n",
    "    # predict the mutants:\n",
    "    # use different prediction strategy by seq length:\n",
    "    if seq_len == optimal_x_length:\n",
    "        x = prepare_x(all_mutants, model_x_length)\n",
    "        Model = load_model(model)\n",
    "        y_pred = Model.predict(x)\n",
    "    else:\n",
    "        y_pred = predict_big(big_seqs=all_mutants, k=optimal_x_length, model_x_length=model_x_length,\n",
    "                             model=model)\n",
    "    # convert array to dataframe:\n",
    "    pred_df = pd.DataFrame(y_pred, columns = ['nuc','cyt'])\n",
    "    # calculate values:\n",
    "    pred_df['nuc'] = 2**pred_df['nuc']\n",
    "    pred_df['cyt'] = 2**pred_df['cyt']\n",
    "    pred_df['expression'] = pred_df['cyt']*(16/17) + pred_df['nuc']*(1/17)\n",
    "    pred_df['export'] = pred_df['cyt']/pred_df['nuc']\n",
    "    # calculate delta:\n",
    "    values = pred_df[target].values\n",
    "    deltas = values[0] - values[1:]\n",
    "    # median deltas of each original nucleotide/position:\n",
    "    delta_medians = []\n",
    "    for i in range(3, len(deltas)+1, 3):\n",
    "      median = np.median(deltas[i-3:i])\n",
    "      delta_medians.append(median)\n",
    "    # form the final data.frame suitable for logomaker:\n",
    "    seq_list = list(seq)\n",
    "    df = pd.DataFrame(columns=['A', 'C', 'G', 'T'])\n",
    "    for i, letter in enumerate(seq_list):\n",
    "      df.at[i, letter] = delta_medians[i]\n",
    "    df = df.fillna(0)\n",
    "    df = df.astype('float64')\n",
    "    # change row index to 1,2,3...:\n",
    "    df.index = range(1, len(df) + 1)\n",
    "    #\n",
    "    return df\n",
    "\n",
    "\n",
    "# Function that performs one evolutionary step on the context of a k-mer:\n",
    "def evolve_target_saliency_once(seq, left_i=19, right_i=26, target='expression',\n",
    "                                 negative=True, decreasing=False):\n",
    "    # generate all mutants:\n",
    "    all_mutants = all_possible_mutations(seq)\n",
    "    # remove mutations on the k-mer:\n",
    "    del all_mutants[(left_i-1)*3 : right_i*3]\n",
    "    # add the original seq to the mutant list:\n",
    "    all_mutants.insert(0, seq)\n",
    "    # calculate target saliency of each mutant:\n",
    "    mut_saliencies = pd.DataFrame(columns=['seq', 'saliency'])\n",
    "    for the_seq in all_mutants:\n",
    "        the_N45_saliency_df = saliency_df(the_seq, target=target)\n",
    "        the_targeted_saliency = the_N45_saliency_df.loc[left_i:right_i].sum().sum()\n",
    "        if negative == True:\n",
    "            the_targeted_saliency = -the_targeted_saliency\n",
    "        new_data = pd.DataFrame({'seq': [the_seq], 'saliency': [the_targeted_saliency]})\n",
    "        mut_saliencies = pd.concat([mut_saliencies, new_data], ignore_index=True)\n",
    "    # only report improved ones:\n",
    "    if decreasing==True:\n",
    "        mut_saliencies = mut_saliencies.loc[mut_saliencies['saliency']<mut_saliencies['saliency'][0]]\n",
    "    else:\n",
    "        mut_saliencies = mut_saliencies.loc[mut_saliencies['saliency']>mut_saliencies['saliency'][0]]\n",
    "    # sort df by target:\n",
    "    mut_saliencies = mut_saliencies.sort_values(by='saliency', ascending=decreasing)\n",
    "    #\n",
    "    return mut_saliencies\n",
    "\n",
    "\n",
    "# Function that evolves the context of a k-mer to the extreme:\n",
    "def extreme_target_saliency_evolve(seq, left_i=19, right_i=26, target='expression',\n",
    "                                   negative=True, decreasing=False):\n",
    "    # initiate the evolution:\n",
    "    df = evolve_target_saliency_once(seq=seq, left_i=left_i, right_i=right_i, target=target,\n",
    "                                     negative=negative, decreasing=decreasing)\n",
    "    # continue evolution if possible:\n",
    "    if len(df.index)>0:\n",
    "        new_seq = df.iloc[0]['seq']\n",
    "        new_df = df\n",
    "        while len(new_df.index)>0:\n",
    "            new_df = evolve_target_saliency_once(seq=new_seq, left_i=left_i, right_i=right_i, target=target,\n",
    "                                     negative=negative, decreasing=decreasing)\n",
    "            if len(new_df.index)>0:\n",
    "                new_seq = new_df.iloc[0]['seq']\n",
    "                df = new_df\n",
    "    return df\n",
    "\n",
    "\n",
    "# function to calculate the connections between nucleotides:\n",
    "import numpy as np\n",
    "import matplotlib.pylab as plt\n",
    "def connections_array (seq, target='expression', optimal_x_length=optimal_x_length,\n",
    "                     model_x_length=model_x_length, model=model):\n",
    "    # generate all point mutations:\n",
    "    all_mutants = all_possible_mutations(seq)\n",
    "    # add the original seq to the mutant list:\n",
    "    all_mutants.insert(0, seq)\n",
    "    # generate a list of saliency data.frames for each mutant:\n",
    "    df_list = [saliency_df(seq=seq, optimal_x_length=optimal_x_length, model_x_length=model_x_length,\n",
    "                       model=model, target=target) for seq in all_mutants]\n",
    "    # convert the data.frames into arrays by summing ATCG values:\n",
    "    all_saliencies = np.vstack([df.sum(axis=1) for df in df_list])\n",
    "    # calculate fold changes:\n",
    "    all_saliencies_fc = np.divide(all_saliencies[1:], all_saliencies[0])\n",
    "    # group the mutants by the position of mutations:\n",
    "    sub_arrays = np.array_split(all_saliencies_fc, len(all_saliencies_fc)/3)\n",
    "    # take medians:\n",
    "    medians_array = np.array([np.median(sub_array, axis=0) for sub_array in sub_arrays])\n",
    "    #\n",
    "    return medians_array\n",
    "\n",
    "\n",
    "# function to plot the connections:\n",
    "import seaborn as sns\n",
    "def plot_connections (array, size):\n",
    "    fig, ax = plt.subplots(figsize=(size, size))\n",
    "    ax = sns.heatmap(array, linewidth=0, center=1, cbar_kws={\"shrink\": .5}, vmin=0, vmax=2)\n",
    "    ax.set_aspect('equal')\n",
    "    ax.invert_yaxis()\n",
    "    plt.ylabel('Mutated position', fontsize = 15, weight='bold')\n",
    "    plt.xlabel('Affected position', fontsize = 15, weight='bold')\n",
    "    #\n",
    "    ax.set_xticks(np.linspace(1.5, array.shape[1]-1.5, num=int(array.shape[1]/2)))\n",
    "    ax.set_yticks(np.linspace(1.5, array.shape[0]-1.5, num=int(array.shape[0]/2)))\n",
    "    #\n",
    "    ax.set_xticklabels([num for num in range(2, array.shape[1]+1, 2)])\n",
    "    ax.set_yticklabels([num for num in range(2, array.shape[0]+1, 2)])\n",
    "    plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "07VnwnlFWhiB",
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "# k-mer context evolve:\n",
    "the_extreme_context_df = extreme_target_saliency_evolve(\n",
    "    seq='TTTTTTTTTTTTTTTTTTCTCCTCAATTTTTTTTTTTTTTTTTTT', # TTTTTTTTTTTTTTTTTTAAAAAAAATTTTTTTTTTTTTTTTTTT\n",
    "    left_i=19, right_i=26, target='export', negative=False, decreasing=True)\n",
    "the_extreme_context_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "P9wjQjBnWhiC"
   },
   "outputs": [],
   "source": [
    "the_extreme_context_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "N1Mydv6HWhiC",
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "N45 = 'CACGCTCGTAATACGGTGAAAAAAAAAAACCCCTTCTCCCTTCCC'\n",
    "the_N45_saliency_df = saliency_df(seq=N45, target='expression')\n",
    "the_N45_saliency_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "hbRIPTUNWhiC"
   },
   "outputs": [],
   "source": [
    "plot_saliency(df=the_N45_saliency_df, negative=True, figsize=[8,2], xticks=True, yticks=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "MEM1i9daWhiD"
   },
   "outputs": [],
   "source": [
    "plot_saliency(df=the_N45_saliency_df, negative=False, figsize=[8,2], xticks=True, yticks=True,\n",
    "             ylim=[-0.04, 0.09])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "D0PZXmDuWhiD"
   },
   "outputs": [],
   "source": [
    "con_arr = connections_array (seq=N45, target='export')\n",
    "con_arr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "kHIh5ItEWhiD"
   },
   "outputs": [],
   "source": [
    "plot_connections(array=con_arr, size=9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "GiwA6PIVWhiD"
   },
   "outputs": [],
   "source": [
    "# quantify k-mer context effects:\n",
    "\n",
    "# function to generate random sequences containing a fixed k-mer in certain position:\n",
    "import random\n",
    "def kmer_random_contexts(num_seq=8, seq_len=45, fixed_kmer=\"TTTTTTTT\", fixed_kmer_start=19):\n",
    "    sequences = []\n",
    "    for _ in range(num_seq):\n",
    "        random_seq = ''.join(random.choice('ATCG') for _ in range(seq_len))\n",
    "        sequence = random_seq[:(fixed_kmer_start-1)] + fixed_kmer + random_seq[(fixed_kmer_start-1) + len(fixed_kmer):]\n",
    "        sequences.append(sequence)\n",
    "    return sequences\n",
    "\n",
    "kmer = 'TATGTTTT'\n",
    "phenotype = 'export'\n",
    "seq_len = 45\n",
    "kmer_position = 19\n",
    "num_seq = 256\n",
    "\n",
    "seqs = kmer_random_contexts(num_seq=num_seq, seq_len=seq_len, fixed_kmer=kmer, fixed_kmer_start=kmer_position)\n",
    "seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "IiLK7k6DWhiE",
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "# negative control:\n",
    "kmer = 'Random'\n",
    "phenotype = 'export'\n",
    "seq_len = 45\n",
    "kmer_position = 19\n",
    "num_seq = 256\n",
    "\n",
    "# generate pure random seqs:\n",
    "import random\n",
    "def generate_random_sequence(length):\n",
    "    sequence = ''\n",
    "    nucleotides = ['A', 'T', 'C', 'G']\n",
    "    for _ in range(length):\n",
    "        sequence += random.choice(nucleotides)\n",
    "    return sequence\n",
    "\n",
    "seqs = []\n",
    "for _ in range(num_seq):\n",
    "    sequence = generate_random_sequence(seq_len)\n",
    "    seqs.append(sequence)\n",
    "\n",
    "seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "E7ViqxlDWhiE"
   },
   "outputs": [],
   "source": [
    "# calculate k-mer position index:\n",
    "left_i = kmer_position\n",
    "right_i = left_i+len(kmer)-1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "TQZRf-OCWhiE"
   },
   "outputs": [],
   "source": [
    "# for NC:\n",
    "kmer_len = 8\n",
    "left_i = kmer_position\n",
    "right_i = left_i+kmer_len-1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "sUVvlnt6WhiE"
   },
   "outputs": [],
   "source": [
    "# create a new df:\n",
    "kmer_saliencies = pd.DataFrame(columns=['kmer', 'position', 'seq', 'phenotype', 'saliency'])\n",
    "# run saliency tests:\n",
    "for the_seq in seqs:\n",
    "    # generate saliency df:\n",
    "    the_seq_saliency_df = saliency_df(the_seq, target=phenotype)\n",
    "    # calculate k-mer saliency:\n",
    "    the_targeted_saliency = the_seq_saliency_df.loc[left_i:right_i].sum().sum()\n",
    "    new_data = pd.DataFrame({'kmer': [kmer],\n",
    "                             'position': [kmer_position],\n",
    "                             'seq': [the_seq],\n",
    "                             'phenotype': [phenotype],\n",
    "                             'saliency': [the_targeted_saliency]})\n",
    "    kmer_saliencies = pd.concat([kmer_saliencies, new_data], ignore_index=True)\n",
    "kmer_saliencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "q98G3xtBWhiF"
   },
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "sns.boxplot(data=kmer_saliencies, x=\"saliency\", y=\"kmer\", width=.5)\n",
    "plt.axvline(x=0, color='black', linestyle='--')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "OU-5nJkBWhiF"
   },
   "outputs": [],
   "source": [
    "# add new results to the df:\n",
    "all_kmer_saliencies = all_kmer_saliencies.append(kmer_saliencies, ignore_index=True)\n",
    "all_kmer_saliencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "-fJWSePzWhiF"
   },
   "outputs": [],
   "source": [
    "# saliency medians:\n",
    "group_medians = all_kmer_saliencies.groupby('kmer')['saliency'].median()\n",
    "group_medians"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "BUPHdTfQWhiF"
   },
   "outputs": [],
   "source": [
    "# normalize saliencies:\n",
    "all_kmer_saliencies['norm_saliency'] = all_kmer_saliencies.apply(\n",
    "    lambda row: row['saliency'] / group_medians[row['kmer']], axis=1\n",
    ")\n",
    "all_kmer_saliencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "mByst95dWhiG"
   },
   "outputs": [],
   "source": [
    "all_kmer_saliencies[all_kmer_saliencies['kmer'] == 'TATTTATT']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "ENnW7mcWWhiG"
   },
   "outputs": [],
   "source": [
    "# draw boxplot as final result:\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "kmer_order = [\n",
    "    'AGGTAAGT', 'TTTTTTTT', 'AAAAAAAA', 'GGGGGGGG', 'TATGTTTT',\n",
    "    'TCGTCCCG', 'CTCCTCAA', 'ACGCCAGT', 'AACCACGT',\n",
    "    'TGAAGAAA', 'ACCCAGAA',\n",
    "    'TATTTATT', 'Random'\n",
    "]\n",
    "fig = plt.figure(figsize=(6, 6))\n",
    "sns.boxplot(data=all_kmer_saliencies, x=\"norm_saliency\", y=\"kmer\", order=kmer_order, width=.5)\n",
    "ax = plt.gca()\n",
    "ax.set_xlabel('Fold changes in regulatory correlation', fontsize=13)\n",
    "ax.set_ylabel('Regulatory k-mers', fontsize=13)\n",
    "plt.xlim(-2, 4)\n",
    "plt.axvline(x=1, color='red', linestyle='--')\n",
    "plt.axvline(x=0, color='black', linestyle='--')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "EAZkPwFwWhiG"
   },
   "outputs": [],
   "source": [
    "# save:\n",
    "all_kmer_saliencies.to_csv('kmer_random_context_saliencies.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "id": "r3q8xR_AWhiH"
   },
   "outputs": [],
   "source": [
    "# Function that performs one evolutionary step on the context of a k-mer:\n",
    "def evolve_target_saliency_once(seq, left_i=19, right_i=26, target='expression',\n",
    "                                 negative=True, decreasing=False):\n",
    "    # generate all mutants:\n",
    "    all_mutants = all_possible_mutations(seq)\n",
    "    # remove mutations on the k-mer:\n",
    "    del all_mutants[(left_i-1)*3 : right_i*3]\n",
    "    # add the original seq to the mutant list:\n",
    "    all_mutants.insert(0, seq)\n",
    "    # calculate target saliency of each mutant:\n",
    "    mut_saliencies = pd.DataFrame(columns=['seq', 'saliency'])\n",
    "    for the_seq in all_mutants:\n",
    "        the_N45_saliency_df = saliency_df(the_seq, target=target)\n",
    "        the_targeted_saliency = the_N45_saliency_df.loc[left_i:right_i].sum().sum()\n",
    "        if negative == True:\n",
    "            the_targeted_saliency = -the_targeted_saliency\n",
    "        mut_saliencies = mut_saliencies.append({'seq': the_seq, 'saliency': the_targeted_saliency},\n",
    "                                               ignore_index=True)\n",
    "    # only report improved ones:\n",
    "    if decreasing==True:\n",
    "        mut_saliencies = mut_saliencies.loc[mut_saliencies['saliency']<mut_saliencies['saliency'][0]]\n",
    "    else:\n",
    "        mut_saliencies = mut_saliencies.loc[mut_saliencies['saliency']>mut_saliencies['saliency'][0]]\n",
    "    # sort df by target:\n",
    "    mut_saliencies = mut_saliencies.sort_values(by='saliency', ascending=decreasing)\n",
    "    #\n",
    "    return mut_saliencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "id": "BsSEdv92WhiH"
   },
   "outputs": [],
   "source": [
    "# function to make a single-step evolution:\n",
    "def evolve_once (seq, target='expression', decreasing=False, exclude=['QWERT']):\n",
    "    # generate all mutants:\n",
    "    all_mutants = all_possible_mutations(seq)\n",
    "    # add the original seq to the mutant list:\n",
    "    all_mutants.insert(0, seq)\n",
    "    # check seq length:\n",
    "    seq_len = len(seq)\n",
    "    # predict the mutants:\n",
    "    # use different prediction strategy by seq length:\n",
    "    if seq_len == optimal_x_length:\n",
    "        x = prepare_x(all_mutants, model_x_length)\n",
    "        Model = load_model(model)\n",
    "        y_pred = Model.predict(x)\n",
    "    else:\n",
    "        y_pred = predict_big(big_seqs=all_mutants, k=optimal_x_length)\n",
    "    # convert array to dataframe:\n",
    "    pred_df = pd.DataFrame(y_pred, columns = ['nuc','cyt'])\n",
    "    # calculate values:\n",
    "    pred_df['nuc'] = 2**pred_df['nuc']\n",
    "    pred_df['cyt'] = 2**pred_df['cyt']\n",
    "    pred_df['expression'] = pred_df['cyt']*(16/17) + pred_df['nuc']*(1/17)\n",
    "    pred_df['export'] = pred_df['cyt']/pred_df['nuc']\n",
    "    # add seq to df:\n",
    "    pred_df['seq'] = all_mutants\n",
    "    # only report improved ones:\n",
    "    if decreasing==True:\n",
    "        pred_df = pred_df.loc[pred_df[target]<pred_df[target][0]]\n",
    "    else:\n",
    "        pred_df = pred_df.loc[pred_df[target]>pred_df[target][0]]\n",
    "    # exclude certain motifs:\n",
    "    rows_to_remove = pred_df[pred_df['seq'].str.contains('|'.join(exclude))]\n",
    "    pred_df = pred_df[~pred_df['seq'].isin(rows_to_remove['seq'])]\n",
    "    # sort df by target:\n",
    "    pred_df = pred_df.sort_values(by=target, ascending=decreasing)\n",
    "    #\n",
    "    return pred_df\n",
    "\n",
    "\n",
    "# Function to make extreme evolutions:\n",
    "def extreme_evolve(seq, target='expression', decreasing=False, exclude=['QWERT']):\n",
    "    # initiate the evolution:\n",
    "    df = evolve_once(seq, target=target, decreasing=decreasing, exclude=exclude)\n",
    "    # continue evolution if possible:\n",
    "    if len(df.index)>0:\n",
    "        new_seq = df.iloc[0]['seq']\n",
    "        new_df = df\n",
    "        while len(new_df.index)>0:\n",
    "            new_df = evolve_once(new_seq, target=target, decreasing=decreasing, exclude=exclude)\n",
    "            if len(new_df.index)>0:\n",
    "                new_seq = new_df.iloc[0]['seq']\n",
    "                df = new_df\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "ejVitxsLWhiH",
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "# generate random sequences without certain elements:\n",
    "num_sequences = int(1e4)\n",
    "sequence_length = 360\n",
    "elements_to_replace = ['GT', 'GGGG', 'AAAA', 'TTTT']\n",
    "\n",
    "# function to generate a random seq:\n",
    "import random\n",
    "def generate_random_sequence(length):\n",
    "    bases = ['A', 'C', 'G', 'T']\n",
    "    return ''.join(random.choices(bases, k=length))\n",
    "\n",
    "# function to replace all elements in a seq to random:\n",
    "import re\n",
    "import random\n",
    "def replace_elements(sequence, elements):\n",
    "    pattern = r'|'.join([re.escape(e) for e in elements])\n",
    "    while True:\n",
    "        matches = re.findall(pattern, sequence)\n",
    "        if not matches:\n",
    "            break\n",
    "        for match in matches:\n",
    "            replacement = ''.join(random.choice('ACGT') for _ in range(len(match)))\n",
    "            sequence = sequence.replace(match, replacement, 1)\n",
    "    return sequence\n",
    "\n",
    "# perform generation:\n",
    "sequences = []\n",
    "for _ in range(num_sequences):\n",
    "    sequence = generate_random_sequence(sequence_length)\n",
    "    sequence = replace_elements(sequence, elements_to_replace)\n",
    "    sequences.append(sequence)\n",
    "\n",
    "sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "H5dxZ6LnWhiI"
   },
   "outputs": [],
   "source": [
    "# or:\n",
    "sequences = ['CATTCTCCACTTCTTGTTCCCCACTGACAGCCTCCCACCCCCATCTCTCCCTCCCCTGCCATTTTGGGTTTTGGGTCTTTGAACCCTTGCTTGCAATAGGTGTGCGTCAGAAGCACCCAGGACTTCCATTTGCTTTGTCCCGGGGCTCCACTGAACAAGTTGGCCTGCACTGGTGTTTTGTTGTGGGGAGGAGGATGGGGAGTAGGACATACCAGCTTAGATTTTAAGGTTTTTACTGTGAGGGATGTTTGGGAGATGTAAGAAATGTTCTTGCAGTTAAGGGTTAGTTTACAATCAGCCACATTCTAGGTAGGGGCCCACTTCACCGTACTAACCAGGGAAGCTGTCCCTCACTGTTGAATTTTCTCTAACTTCAAGGCCCATATCTGTGAAATGCTGGCATTTGCACCTACCTCACAGAGTGCATTGTGAGGGTTAATGAAATAATGTACATCTGGCCTTGAAACCACCTTTTATTACATGGGGTCTAGAACTTGACCCCCTTGAGGGTGCTTGTTCCCTCTCCCTGTTGGTCGGTGGGTTGGTAGTTTCTACAGTTGGGCAGCTGGTTAGGTAGAGGGAGTTGTCAAGTCTCTGCTGGCCCAGCCAAACCCTGTCTGACAACCTCTTGGTGAACCTTAGTACCTAAAAGGAAATCTCACCCCATCCCACACCCTGGAGGATTTCATCTCTTGTATATGATGATCTGGATCCACCAAGACTTGTTTTATGCTCAGGGTCAATTTCTTTTTTCTTTTTTTTTTTTTTTTTTCTTTTTCTTTGAGACTGGGTCTCGCTTTGTTGCCCAGGCTGGAGTGGAGTGGCGTGATCTTGGCTTACTGCAGCCTTTGCCTCCCCGGCTCGAGCAGTCCTGCCTCAGCCTCCGGAGTAGCTGGGACCACAGGTTCATGCCACCATGGCCAGCCAACTTTTGCATGTTTTGTAGAGATGGGGTCTCACAGTGTTGCCCAGGCTGGTCTCAAACTCCTGGGCTCAGGCGATCCACCTGTCTCAGCCTCCCAGAGTGCTGGGATTACAATTGTGAGCCACCACGTCCAGCTGGAAGGGTCAACATCTTTTACATTCTGCAAGCACATCTGCATTTTCACCCCACCCTTCCCCTCCTTCTCCCTTTTTATATCCCATTTTTATATCGATCTCTTATTTTACAATAAAACTTTGCTGCCA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "7HScBE3kWhiI",
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "# check elements:\n",
    "elements_to_check = elements_to_replace\n",
    "#\n",
    "def check_sequences(dna_sequences, elements):\n",
    "    containing_sequences = []\n",
    "    for index, sequence in enumerate(dna_sequences):\n",
    "        for element in elements:\n",
    "            if element in sequence:\n",
    "                containing_sequences.append(index)\n",
    "                break\n",
    "    return containing_sequences\n",
    "#\n",
    "check_sequences(sequences, elements_to_check)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "id": "vj3IHeoyWhiI"
   },
   "outputs": [],
   "source": [
    "# define model .hdf5 file:\n",
    "model = '/content/L5-220528_em5-LSTM64x32x0.5-64x0.5-rep4.hdf5'\n",
    "# em5>LSTM64x32*0.5>64*0.5-rep4\n",
    "# CNN8*512>GAP>128\n",
    "# CNN8*256>GAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "LrYJHGP1WhiI"
   },
   "outputs": [],
   "source": [
    "# apply prediction:\n",
    "y_pred = predict_big(sequences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "c6BQW9p7WhiJ"
   },
   "outputs": [],
   "source": [
    "y_pred"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "7c6boUNwWhiJ"
   },
   "outputs": [],
   "source": [
    "# convert result:\n",
    "pred_df = pd.DataFrame(y_pred, columns = ['nuc','cyt'])\n",
    "# calculate values:\n",
    "pred_df['nuc'] = 2**pred_df['nuc']\n",
    "pred_df['cyt'] = 2**pred_df['cyt']\n",
    "pred_df['expression'] = pred_df['cyt']*(16/17) + pred_df['nuc']*(1/17)\n",
    "pred_df['export'] = pred_df['cyt']/pred_df['nuc']\n",
    "pred_df['seq'] = sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "cQ6affqmWhiK"
   },
   "outputs": [],
   "source": [
    "# sort:\n",
    "pred_df = pred_df.sort_values(by='export', ascending=True)\n",
    "pred_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "c0YWYGILWhiK"
   },
   "outputs": [],
   "source": [
    "# save:\n",
    "pred_df.to_csv('N360_nonGT,GGGG,AAAA,TTTT_1e4.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "X_o43ftlWhiL"
   },
   "outputs": [],
   "source": [
    "# load:\n",
    "import pandas as pd\n",
    "pred_df = pd.read_csv('N360_nonGT_1e3.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "13cPhlcbWhiL"
   },
   "outputs": [],
   "source": [
    "pred_df.iloc[0]['seq']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "id": "1qV3O-uAWhiL"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "81"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 3'UTR saliency test:\n",
    "# define 3'UTR sequence (ATCG only):\n",
    "seq = 'GCGCAATTCCCCATTCTCCAAGGGCTGAGAACTTCGGTGAGTTCATCCACCTGTCTATTTGCACATGGAGGACTTTTCAAG'\n",
    "len(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "id": "ySu3wYqXWhiL"
   },
   "outputs": [],
   "source": [
    "# define target:\n",
    "target = 'expression'  # 'expression'/'export'\n",
    "negative = False  # True/False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "id": "F4wIVi7eWhiM",
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "283/283 [==============================] - 2s 5ms/step\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1607171/4129292667.py:101: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  df = df.fillna(0)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.003331</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000436</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.006959</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003412</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>-0.000122</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003633</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>0.001557</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>0.000971</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>81 rows  4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           A         C         G         T\n",
       "1   0.000000  0.000000 -0.003331  0.000000\n",
       "2   0.000000  0.000436  0.000000  0.000000\n",
       "3   0.000000  0.000000 -0.006959  0.000000\n",
       "4   0.000000  0.003412  0.000000  0.000000\n",
       "5  -0.000122  0.000000  0.000000  0.000000\n",
       "..       ...       ...       ...       ...\n",
       "77  0.000000  0.000000  0.000000 -0.000268\n",
       "78  0.000000  0.003633  0.000000  0.000000\n",
       "79  0.001557  0.000000  0.000000  0.000000\n",
       "80  0.000971  0.000000  0.000000  0.000000\n",
       "81  0.000000  0.000000  0.000080  0.000000\n",
       "\n",
       "[81 rows x 4 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# nucleotide saliency test:\n",
    "sal_df = saliency_df(seq=seq, target=target)\n",
    "sal_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "JCy2a6EAXOEg"
   },
   "outputs": [],
   "source": [
    "# save:\n",
    "sal_df.to_csv('/content/TP53-3UTR-expression-saliency.csv', index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0YAAADUCAYAAAC4YjPNAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjkuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8hTgPZAAAACXBIWXMAAA9hAAAPYQGoP6dpAAA7tUlEQVR4nO3deXyU5b3///csySQhJCGQEBYhLLIpyiYIWjdQAddqtR6xVg9Fz+8UrcupB9pqta0/vm21ekSstdrWfitHaz0qdcGDIqKIhE0FZEdBlgAhZN9m+/5xZTIzmSUzkxlCzOv5eMwj99xzzdxXQsg97/lc13VbvF6vVwAAAADQhVk7ugMAAAAA0NEIRgAAAAC6PIIRAAAAgC6PYAQAAACgyyMYAQAAAOjyCEYAAAAAujyCEQAAAIAuj2AEAAAAoMsjGAEAAADo8ghGANBB1qyRzjlHsliCbxMmSIcOdXTvAADoWuwd3QEA6Io+/VSaOlWqrQ19bP16ac8eqU+fE94tAAC6LIIRAJxgXq/0b//mD0Xdukk//KFUWCh98IH0z392bP8AAOiKLF6v19vRnQCArmTrVmnUKLOdlyetWycNGeJ//J13TLXojDNS14eGBunYMSk7W8rNTd1xAADoLJhjBAAn2Lvv+rfvuCM4FEnSpZemJhR5vdKSJdKVV5pA1L+/CWannio9/HDyjwcAQGdCxQgATrBvf1t67TWzvWKFdP75qT+m1yv96EfSwoXhH+/WTaqpSX0/AAA4WVExAoATbO9e//bYsSfmmEuW+ENR797S4sVmjlNlpXls8uQT0w8AAE5WVIwA4AQbNkzaudNsezxmie5Uu/JK/6IOy5dLF14Y/LjXe2L68U1VUmJ+vh98IB05ItntUnGxqQbOnStlZnZ0DwEAbSEYAcAJNmaM9NlnZru2VsrKSu3xamvNXCKXS+rXT/r6a0JQsjid0u23S3/+s7mfl2eqgN26STt2mNvevdKAAck97vHjpur3wQdm6feaGnPMIUNMCP73f4/v9eqd9br7nbvl8XqU48jRby/+rSwnwy9JU4V09COpfJ3kqpFsmVK3QVKvyVLuyI7uHZCwryu/Vp2zTpI0rOewk+P/G1iuG0CMvF7JWSE5qyR7Nyk9X7KEGY3r9UquasnrkdJywreJJpaTQyf/PKdvX38w2rRJmjQptcc7etSEIkkaOZJQlEzPPecPRddeK/3pT1JOjv/xzZulHj2Se8zVq6UrrjCrCjoc0ne+Y6qQHo853l/+En8wemHTC/rD+j+03J8xdIamDp6a3I7Hw+uVtvzK3DxNkqNAyh4ied1S9XbJYpOuLe+4/uEbpXWNINUh5Xj9cY16apRqmszEzqWzlurSoZem9JiIDcEIQGRej/Tl89KXf5XKVkkel5TeQ3LXSbJIg2dLExZK9YelXU9J+/4h1ew2b2Ss6eZNTPeh0tl/lXqeFeMxA05QgSenTh6GAl1wgfT222b7gw9SH4wCh3GxwEJyPfaY+WqxSM8+GxyKJOn005N7PKdTuvVW/1Lrn3winXZacJumpvhe0+v16rFPzDdy/WnX6+9b/q7HPnmsY4PR169Imx4w22N/Jw27Q7I2v2XxuKXj6zuub+gcqndJX/+PVLlJajgquevNh3XdBkoDb5AKzm1petniy/T2LvNH+b4p9+nXF/868eO6G6SGI1JTuTmHpveQMvtJtvSWJovWLlJNU40G9xisPcf3aMFHCwhGJwmG0gGI7NN50tZfS/ZsaeKzUv+rJZvDhJS6r6W6fVLeGdLbZ0i1e6WC86Rxj5l9VrvUVCkd3yh1K5ayi+M//okKRl6vVL1TKvvYnNCclWZ/Wq7UbYDU7wpTJUuSDRuk8ePNdr9+pnrUs6f/8U2bpPR0afjwGF4shk82PW6vBg2S9u0zc1+OHjVDvtB+3bpJdXXm36+srO32ja5GldWZhtnp2crNiO8iUmvWSGefbbavuUZ65ZV4exzqnV3vaPoL0zU0f6g23LZB/R/rr6rGKm394VaN6DWi/QdIxHsXSUfel3JGSZdtMfvKN0pVW/1tCi+Qsvp2SPdwktvy/0uf/9R8QDf8HqnwPCk9z5yTqrZKWf2kAddLkt7e+bZmLp6p0wpO08Hqg6p11uqLf/9CQ/KHRD9Gawf+KW1/Qjq6UsrqL+WMkKwOqbFMqt4hnfuKVHCOaptqNfDxgTpWf0wbb9+oK//7Sn1d9bVW/esqTTllStBL1tRIX34pVVWZ7fR0qVcvM3+xe/fk/KgQ7ISsSrdo0SIVFxcrIyNDkyZNUklJSdT2L7/8skaMGKGMjAyNHj1ab7311onoJtD1eJxS1XbpWIl0+H3p8Arp2Fqp5kvJ7ZS2/860K/6eNPC7JhTtWChtuFva9jtTIdrzFxOKJOms30v546TGoyZUfbFAOrRUOvZJR32Hbas/JL0zXnpzuKmOZRZJfS6R+s6QcoZLtfskZ3VSDzlmjD/0HDhg7v/hD9Krr5oltc86K7Y32ZJMqAu8hdlvtUo33WR2u1zST34SmjNXrGjf99RVnXmm+XrsmH9BjUhcHpeu+fs16v9Yf5268FSNemqU9lftj+t4gW+GKivj7GwEv/vE/D+fPXa20m3puvH0GyVJj3/yeHIOkAhv89hPa8DAltJl0qafS6tnmVvF5x3TN3Qcj1Pa82dp9c3SOxOllVdLq/5F+vgmadUN0qrvSjVfSZ//zLQf/QtpzALz97xuv9R0TMoolCxpkiSn26l7//deSdJvLv6N5p07T03uJt337n3x9evIh9KHV0uH35XG/Zd0+S7pW0uks5+Xzn9TunyH+cBQ0rMbntWx+mOaNniaxhSN0dyJcyVJCz5a0PJyf/qTOS/k5kp33iktXSp98YW0fr30/PNcdy6VUl4xeumll3TzzTfr6aef1qRJk/T444/r5Zdf1vbt21VYWBjS/uOPP9Z5552nBQsW6PLLL9fixYv161//Whs2bNDpyR6TAHRVFZuldf+fdGyNNOC7UtElZgy/xWrmB9Xtl4pvkt47X6rcYoYcXPS+eZNy4J8mRG1vHkP0rVelD79tts95yXwK11gm7X1J2v2MefMycJY05W/x9/NEVIw+/U9p62/M9neqpbRsqWyNf58kDbjODL1Iovffly6+WHK7wz/+0UfSOeck8MK+n1mrn9fBg2ZRgCNHzP2RI6WrrjJDs1auNCddhtnFb+lSacYMsz16tPTHP5pga7Waytz//I+Ze9Szp1d3vH2HFq1dpLsm3aVLhlyiyxZfpjOLztTKW1aquyP2j38vv1x6802z/dRT0m23STabue9ySRs3mj7EYvORzRr9+9FhH8u0Z2rf3fvUK6tXzH1Lmj1/kdbcarYnLzb//ywWyVUrvZxt9p//ttR3+onvm6+Lx/dob4X5UKhfTj8N6zmsw/rSZZTMkXY/K2X2kS5ZZyqGFZulqi/M416P1H2YtGyK5GmURv1EOrM5RXz+gHT0Q+nIClPRueprPVnypO54+w71yOihn533M9U763X/+/fLK6/e//77uqD4gtj6tfMpad0PzfZlW021qHKrtPZ2qWKTmZ87eLaaJjylIU8M0f6q/eqR0UM9MnuowdWgg9UHTRf/7XOVfTFaF11kXmrWLOlvCZw6kbiUB6NJkybprLPO0pNPPilJ8ng8OuWUU3THHXdo3rx5Ie2/+93vqra2Vm+88UbLvrPPPltjxozR008/ncquJo3vg1rfGx6r1X/S6sy2b5cef1zats2UcS+5xIyn931vHo8ZFtStmxkqVFYmFRVJhYVSWpr5OXi9pt3o0aZdrBpcDdpwaIOcbqe88qpf934amj+UVVwStfs5qeQHkizShe+aYQbVO824/l2/N8Hm9AekPjOkj64xFaG8MVL/K02AqvjMnJwkMwF6629NkLDYTdDKO0OyZUg7n5SqtkkDZ2lrj79p7Vrp8GHplFOkPn3MsADf74XbLZ1xRqvhAUkKRvsq92nHsR2yNi8EMbZorHpkNs+Ir94lLZ9qhgUO/Bczbyqrn3nsgyukml3SafdLZ/xCu8p36eEPH9aKr1boptE36aUtL2lcn3G6/7z7dVrhaRGOHtmrr0p33WWGuAXq10/68ENp0KAEvtkIwUiStmyRbrxR+jzMB+29e0ulpQkc75sq1r8tXq+ee0665x4z3EUyCyJkZZmV4ySzKt2rB/9Ld71zlxw2hy4adJHSbGlae2CtDtUc0sxTZ+r1G16X3RrbtN8jR8w8I99git69paFDzf+hrVvN4h5ffBFb93+w5Ad6buNzunL4lbp1zK0t+3+z6jdavX+1fnXhr/TT834a24slk8ctbbhL2rlIklfKGWnmK7rqpMPvmTYdFIwqGyr10AcPadHaRZo9dray0rK0sGShZo+drV9e+Ev1zOrZ9oskm8dt3vBX7zQfYGUPMSv4eZvfiHjdUmZf8zOMw/6q/fr88OeyW+1ye9waUzRGfbr3CWrjdpvhpE1NZtu39H9amvl/4HD429Y56+TxeiRJ3dK6xX8O3/m0tP6HZojaWX+Qek+VXFVS1Q7pw6tMm2+9ahYJWvtvZr6Pbz5RWq60/zXp639IWf11/JLPNXThUJXXl+t7Z3xPDpvpaGltqd7Y8YbGFo3V2jlrZbPG8AbO3WgqRoeWStlDzZy47sMka5r06Y/NkPLBs/Xn9HP0r0v+Vafmn6ofT/lxy9Pf3PmmXt/+um4cfaN+P+0FXXut9O67UkGB9KtfSd/6llnExe02f6fLy80Hayczr9d88Bbu9yI93QzrPhmlNBg1NTUpKytL//jHP3T11Ve37P/+97+viooKvf766yHPGTBggO655x7dddddLft+/vOf67XXXtNnvmWcAjQ2NqqxsTFon8NTLkfDHrO0Z/ehZm6Au0FqOi7VH5BkkTdvjBrsg9XYaN6oOxzmH8x3a3kth2Sp+1qq/dI8355tftEtVvPmUVYpZ5g+3jxCn3xi3vSNHm1OTBkZJjS4XOaX46yzpKwsrzxej7wyP3aLLC3bVotVVotVlQ2V+uzwZ9pWtk1VjVUamj9UR2uPqs5Zp6LsIo0pGqOh+UOD/7O28w3kzp3S/v1Sfb3Uv78JLdaAgZZut3kj+957Jhg1NJhVkLp3N+0sFvNz7NvXDO9Yt84EowkTTDiy24PfAI8ebb42Nprn+n5Wvm/F6zXtq91HtWzPMq09sFb9c/pr8imTtb9qvzYf2SyLLLpkyCWa1H9S8BuKWH8WUd5ASlJ1tXkDWV5uti0WMy9j4MAwK3tFeK3aWvOzPXLE/A7k5weHRLfbTM72LddcX29WlTp61LypysgwIeL006Vqz2Gt+GqF7Fa7Tu15assxGlwN2l62XX2799W5A87Va9teU6O7URcPvjjoBLbx0EZtOrJJ/XP666JBF5lAUPax+VTL09j8e203Cyuk50t9Z0r54/0n3LLVUsNh8//KN4G1xzgTqixWM7n1wOtSzR4zT8diMSGq+3Cp8Hx9srlYn3xifhZnnml+n1r/LOIKRu4mM267bp/pk7vRrFRlz5Yc+fLmnK7Pjn+lpbuWqrKhUsN7Ddfphadrzf412nN8j4b1HKZLh16q4rxi89yy1WaBiYajZo6RxWYmzXYbKG/vaXq//KD+9vnfVJxXrNGFo5VmS5PL49K2sm3aVrZN1426TtMGT9NH+z7SoZpDGpo/VNnp5pPtJneTdh7bqQx7hi4oviBoXkljo/R//6/5MMFiMQsxfOc7YZbwTtLvtddrhmO8+qo5wWZnm/lOV11lhm2EOPy+WRrZlhH5mM4qM3ylxxmR28TKWWMm1hee37LrUPUhLduzTFaLVTedcVNQ821l21RyoET5mfm6fNjlwa/l9ZhPiANeK2Ft/Fzr6kyY/eADcx6w202wPe88aeBph/TT9+fLYrHotxf/tqUC4/V69cD7D2hv5V7dfObNmjZ4Wlxd2rXLVPs+/dT8jcrKMgHpoovM39jdu81Qzaoq8++cmen/u+52S6cMdOmXG/5dDa4G3XfOfTq90D8qY8VXK/SnjX9Sn+w+LRPRGxtN8Dp40Pxta2w058iePaVRo8ybnYoKsz8nx//Bh+9vutdr3tzF9UFhzVcmCB3fYEJRWnep22DzO9nzrLhXvayrM//XDhww5+b8fPP3vLg4tucfrT2q36z6jcrry3X5sMvlsJs30y6PS2/seENZaVn6z3P+MyQ8pJy70XywVb1TcvSUck83Q5+rtpkRAPWHTJXktJ9IMr8vX3xhzjW1tebNa2am+bccPdqrw54v9NbOt3S07qiG9xyusX3Gat3BddpxbIf65/TXjKEz1L1pmF591aI9e8wHoGefbf6G+P6mNzWZf+/solKVHCjRF0e/kNfr1ejeo7W7fLeqm6o1KG+QJvabGN8HnXUHpCMrzaIKzkpTRbSmSWk9zPd4yrXmwy1ntXTwLTPqoeGwGYaXnmfmvPacqLeOl+vFzS9qcI/BevCCB1te3uVx6Y637lCts1b/MeU/dEav4WZBIVnMcWSR5Gn+pXab30F7VvNc1e3SoXeah+2Vm7bp+VL2IKn3hfrlxpe1s3ynvjPqO7py+JUtx9xzfI8eXPGgbFabFs1cpEx7lrZtM4ur7Nhh/r0C5xhNnCgFvK022vgbFdSmrXbtcPSo9Npr5r1Pbq45p+Xmmr+JvrCUmysNGFKn0ppSldeXq8ndpJ6ZPVXVWNVyyYCi7CLlOnJV66xVZWOlGl2Nslltctgcqmmqkd1qV3dHd+U6cpVmtZu/9y3fo0Vqfm9t9lukWAKuUhyMDh48qH79+unjjz/W5IDLqt9333364IMPtGbNmpDnpKen6/nnn9e//Mu/tOx76qmn9NBDD+nw4cMh7R988EE99NBDQfsmThirccPylJnmUqPLpgaXTV75fhnMVlldZsC+yIrzKjVr9Hb1zq7Tiq/6a3d5rtJsHp2SU61eWQ3KTHPpw719tbE0dFggOqdjx0Zpy5bZqqwcrJycverTZ7UyMsplsbjV1JSjmpr+Ov30Z2Sztb3006FDZ+vYsdPU1JSjoqI1ysg4JqvVLcn8R/V6rerefa8aGnpp27bv6ciR8XK705Sbu0eZmWXyeOyqre0jh6NS55wzX06rU4fzDquse5nya/JVn14vl82lPsf7qEdtD1lkUVl2mdYNWSdJKqwsVGZTpiq6Vag8u1zZDdmauGuiMp2d+2qTM4Z+pe+P2SqvV3pizRjtOZ6r3tm1GtHruK4duUt2q1f/teZMrdzbv0P6V5VZpSM5R5TmTlOaO03VGdXqVd1LPWp6yHpipnYmxdiiI3rowjX65/ZB+uOGyEOZf3jWZzpnwCHNWXKRap3pYdssab667JVXXBH1mHPGbdYVw7/UwysnaM0B8+ayLLtMq4evlrzSFRuCn7+7cLe+OOUL5dbm6rxt5wU9dt7AA/qPKRv0yMdjI/4u+PoVS986k7Ky07R//1Q1NOQrP3+L8vO3yWptan5f5JXXa1VW1mFlZBxv87Xq6/P1+edzdfToGGVmHtWgQW8qO/tr2WwN8njS1NjYQw5HhQoKQj+8bEusvxfxvFa41zt+/FRt3fp9lZefJq/XpvT0CqWnV6upqbuamvI0evRTKi5eGtNrJbNf4drF0ibRflVWDtZnn/1QlZVD1KPHdg0c+I6yskplszXJ7U5TY2O+srIOKS9vT9yvnUzJ/L0AJGnJkiUxtev0wShsxcjhkCOwdgvEyOMxFYu6OjM85cCB1A+DrKgwFZR9+0y17r33TDUuUFVV6DLA0Ww9ulUzF89Ug6tBd068Uw+seEBTB03Vy9e9HNdchpNW+QYzfK/iM/PJcc5IM0xCMp8eehqlkfdxwZ72evd8s8KSLNL0DVKPMaFt3A3Sq0Xmk9vA8fytxfJpZv0h6Z+DzWtm9Jau3CfZ0rVq3yqd+2eztK7zfmdQhfj/fPR/NP+9+Zrcf7I+nv1x+P5n9pOu/LL5094I/Wqrb13Ys89Kc+aY7fvvl37xiyS+eCy/F/G+VqvX27XLVKPr66URI6T//m8zsd1n715T5Qr6u5vM34skVXuT0a/f/U6616w1oAULpPvuCx4dIplK2sk6zAlItZT+6vfq1Us2my0k0Bw+fFhFRUVhn1NUVBRXe0IQkslqNSeNhx82Q75+/nMzcbpfP3OiKCszQwlnzDDDwJJh9Wr/HJPLLgsNRVJ8oUiSRhaM1CezP9FLW16S1+vVUzOf0q1jb415DsNJL3+cdM5/++973JKnQZLVDPkiELVf7d7mUCRJXrMS1Pj/Cm134A3/8uZf/TVyMIrF16+YUCSZoS8H35BOuUZpNv9/Nqc7OBg53U5JCmojyQwV9fW//oBU+q5ZmQpx+8EPzN/AN980i4Zccok0YID/zXNTkxm+d9NN0V+nozzyiAlFkvToo8GhSDLDo1MqmcEq0v4Yj3HPPWZ454oV/jmHvXr5PwBsapKuv1668ML2dRnorFL6Lik9PV3jx4/Xe++91zLHyOPx6L333tPcuXPDPmfy5Ml67733guYYLVu2LKjiBKTSL34h3X67mS/w6afS00/7J1X75hjNSOL7q8mTzZuMffvMG48dO9pfMZKk3tm9deekO5PX0ZOZ1SZZk3edIUg63mpY1KGl4dsd/dC/XbffjOfPjX8hCnPMja2O+bZ0yjVKD7gwYpO7SZlpmUH3JQW1Ma+1IfS1CEYJmzEjuX/3TqTAz059AemkFCncJLmSOWGCuQEIlfKPj++55x59//vf14QJEzRx4kQ9/vjjqq2t1a23mtVvbr75ZvXr108LFpj123/0ox/p/PPP16OPPqrLLrtML774otatW6dnnnkm1V0FWvTrJ91wg7mlWl6etGyZ9NOfmlWmRo40k+H79TOTFHftMhNbV65s86WA5KloFYyqd5hJz74V+3xqdgffP1bSjmD0afD90mWSgkOP0+MMauK7HxKMqrYF3y9bnVif0OnNm2cWODl+3FwnzOEwK3o5HGYI3bp1ZuGBceM6uqcAOlrKg9F3v/tdHT16VA888IBKS0s1ZswYLV26VL1795Yk7du3T9aAAa5TpkzR4sWL9bOf/Uw/+clPdOqpp+q1117jGkb4Rhs2THr5ZfNp5qZNZsje8ePmxH3KKWaFKeCECnfxzMPLpUHfC97XOhiVr5cG36q4eVym2hSodq9UvSukYhQoYsWoanvw/YrPJFe9ZO/cC48gfn36mPDz85+b60755vNnZ5v5pFar9Oc/S+PGJ2/IGoDO6YRMOJg7d27EoXMrwlxu/brrrtN1112X4l4BJ5/MTLMMJ9DhWleMJDNnJzAYeT1SzZfBbVoPYYtVVfOy8SHH/EjpPS9queubU9T6fpsVI4/TLANecG5i/UOnNniwqRpJ0ldfBS/Xfeqp5tII+l60VwDQFXxDZmIDAJLGVRdaCZJC5x3V7Q8NM8c/M4thxHjNCP/zPo24P73QfxHPmCpGvuuJtHashGAEFRdHuG4R1SCgy+s8F9QAAJwYFZv8F8sLVLkleH/1rtA27jqpZmf8x4wUjCo+i22OkTUgGNUfMMu2t1a1Nf5+AQC6DIIRACBY5ebw+92tKknhqkqSVJ1AMKr6Ivz+VsEopopR62F0LfvDVJFw4lks/lu4fSy3D6CDEIwAAMFq90Z+LHBRhpowFSNJqtkT/zHr9off33Rc6Q3+a9u1DkZhV6WLFIAIRt88kcIUIQtAAghGAIBgDaWRH6sMqOxEqhhF2h9N/cGID6XV+0NT68UXfEEp6AKvkQJb4xGpqSL+viG5vN62bwDQAQhGAIBg9YciP+aq829XJykYuRulpvKID1sskt1q1goKqRiFW5UuoMIUgqoRACACghEAIFhDlGAU1C5CZSneoXTRglgzX/BpvfhC2DlGDUcjv1Ai859w8qL6BCCJCEYAgGD1UYbSBWo6Fn5/zZfhV7WLeLzIw+h8fMEnpjlGjVGCkbs+9n4BALoUghEAwM/riT4UzcdZbS6aGo6nUWosi/2YcQSjSHOMgoPRkdiPDQBAM4IRAMCvsUzyumJoF6Fa5ON1x37MGIbuRawYhZtjFE8oAwCgGcEIAOAXS7VIijyMLhExBJmY5xg1VUSuZAEAEAXBCADgF2V1uCBtVYziOubxNpvEPMco2vwiAACiIBgBAPxivc5PUoNR28eMeY5RtBXpAACIgmAEAPCLNRglcyhdDBWjNKu5gGukOUa+x5lfBABIFMEIAODnrIitXTIrRjEcM+Y5RrH2HwCAVghGAAC/k6Fi1GOcZLEF7Yp5jpGzOnn9AgB0KQQjAIBf64pL/gSp+6mh7Zoqk3fMwDDm6CVN+1AacltQk0jBKLRiVJW8fgEAuhSCEQDAr3XFaOIfpbP+ENrOlcTKTGDFqP+3JXuWNOD6oCaRFl8IuY4RwQgAkCB7R3cAAHASCawYZQ+WeoyRvB4pLSc4dLhqknM8j1Ny1/nv551hvuaeFtTMF3weWf2IFq1d1LK/3lUf9HhSAxsAoEshGAEA/AIrRrmnm68WqxlOV77e/1iyglHrCk/e6OZjWoJ2+4JPg6tBDa6GkJehYgQAaC+G0gEA/AIrLq2qNkGStciBu1XIyR0dtllL8IkgYsUo/yxT7QIAoA0EIwCAX2BQiRaMklUxCjxeRpHkyA/brOU6RRGk2Zofb10xmvKCNGp+e3oIAOgiCEYAAD93o387Z1TkdkmbYxRwvKz+EZvFXDEKDEY9xpkhgP2uaE8PAQBdBMEIAOAXY1AJDUaWsM3aFFQx6h2xWezBKGAoXc5I/1dbRmL9AwB0GQQjAICfL6hYbJKjZ/g2HndwoLFlSJfvNNcgSvR4UnKCUeAcI9/1lyxWyZYZf98AAF0KwQgA4OerGDl6mUARTutqUZ+ZUvchUv+rEz+elKRgFLD0d/eh8fcHANBlEYwAAH6+OUaOwshtWgejgnPN156TEjheQMXIURCxWczByNPk3+mrGAEAEIOUBqPy8nLNmjVLOTk5ysvL0+zZs1VTE3nCbnl5ue644w4NHz5cmZmZGjBggO68805VVlamspsAAEnyev0VnIzIISVkSezsQeZr9+HxHzOwYmTPjtgs9mAU8HrZVIwAALFLaTCaNWuWtmzZomXLlumNN97QypUrddttt0Vsf/DgQR08eFCPPPKINm/erL/85S9aunSpZs+encpuAgCk4GqLvXvkdq7a4PvdmoORNYFrhgdWjOzdIjaLFoysFqvsVrvk9Uhed/NOR8SlvwEACCeBs1hstm7dqqVLl2rt2rWaMGGCJGnhwoWaOXOmHnnkEfXt2zfkOaeffrpeeeWVlvtDhgzRww8/rJtuukkul0t2e8q6CwCIMaSEXJTVVzFK4TGjBaOWaxwFBrtIC0cAABBByipGq1evVl5eXksokqRp06bJarVqzZo1Mb9OZWWlcnJyIoaixsZGVVVVBd0aGxvDtgUARBE0rC1KMAoMIOn5UlpOko4ZeShdywVcoz0WeA2mdIIRACA+KQtGpaWlKiwMnrxrt9uVn5+v0tLSmF6jrKxMv/zlL6MOv1uwYIFyc3ODbgsWLGhX3wGgS3LHGoyc/u32VIukpFSMwi68QMUIABCnuIPRvHnzZLFYot62bdvW7o5VVVXpsssu06hRo/Tggw9GbDd//nxVVlYG3ebPn9/u4wNAl+OJcShdYADJDB0WHZegYJTY4gsEIwBAMsQ9aefee+/VLbfcErXN4MGDVVRUpCNHjgTtd7lcKi8vV1FRUdTnV1dXa/r06erevbteffVVpaVFHkLhcDjkcDhi7j8AIILAipEtxmCU1qN9x4xx+F7cwSidhRcAAPGJOxgVFBSooCDKMq7NJk+erIqKCq1fv17jx4+XJC1fvlwej0eTJkW+1kVVVZUuvfRSORwOLVmyRBkZGfF2EQCQiJjnGAUMpUtvZzBK6lC6wGCX1b5+AQC6nJTNMRo5cqSmT5+uOXPmqKSkRKtWrdLcuXN1ww03tKxId+DAAY0YMUIlJSWSTCi65JJLVFtbq+eee05VVVUqLS1VaWmp3G53qroKAJBiX5UuqDKTzGDUzqF07sDlxglGAID4pHT96xdeeEFz587V1KlTZbVade211+qJJ55oedzpdGr79u2qq6uTJG3YsKFlxbqhQ4MvzPfll1+quLg4ld0FgK4t5sUXkhiMfFUei1WyZUZsFvdQuiivBQBAOCkNRvn5+Vq8eHHEx4uLi+X1elvuX3DBBUH3AQAnUEcOpbNlSRZLxGYMpQMApFrKhtIBADqZwGFtMS++kNvOYzaHmWhBTAlUjOxUjAAA8SEYAQCMRCpG7Z3L41sivI2hb2nWKBd49T3GUDoAQDsQjAAARiJzjGztXDnUV6Wy2KI2i23xBYbSAQASRzACABiBF3i1Rg4iSa3MuAMWX4iCxRcAAKlGMAIAGIEVl2gVHG/AULr2VoxaVqULON7mX0jvTAxqFv8cIypGAID4pHRVOgBAJxK4+IIvqHi9Zk5R4IqhgdcLsrYzGHndwceTpIajUu2XQc0Cg5Gtua3H65FXXipGAICkoGIEADACK0G+oHLwTenvDun4hvDt2h1AvMHHiyAwGO2/Z79cD7h099l3Bz8WtFw3wQgAEB+CEQDA8FVvpOhzfpK5+ILX03y82INRU3PFytm8Oh5D6QAAyUAwAgAYgcPlogWVDg5GTrcJRL6A5F+VjqF0AIDEEYwAAM08/s2owSgVQ+liX5WupWLkbl0xYrluAEDiCEYAACPeipE1XbJY2nnM2CpGNqtN1ubw5BtC19TcDy7wCgBIBoIRAKBZYMUo2hyj5opRUsJHbIsvSP4AFLliFBiMHEnoGwCgKyEYAQAMb6xD6ZoDSHvnFwUeM4Zg5AtAEecYBQYjTm8AgDhx5gAAGPEOpUtGMIqjYuQLQBFXpQu6QC2nNwBAfLjAKwCgWQcMpUukYuRpq2Jk8c99qvlKqtwcvGAEAABh8JEaAMCIe/GFJA6li+F01HooXcgcI6/LfA0MdQf/Ka28QnLVtL+vAIBvNIIRAKBZnMt125O4+II1/qF0IRWjluoTpzYAQPw4ewAAjKCKUbShdCmoGCUwlC5kjlEc85UAAGiNYAQAaBZjxaUlGKUl4Zi+MBb7ULqIy3V7Y38tAABa4+wBADC8MVZcfEPpkjJkzXeBWG/UVpKUZjNBrPVy3b79MQc7AADC4OwBAGgWazBqrhglI4C0vEbbwajN5bpjDXYAAIRBMAIAxKdlWeyAU4jHmdiS2L5g5I09GEVcrrsl2HFqAwDEj+sYAQCaNQ9rayukhFsW+4PLpdL/TfyYgSviRdD2HKPYl/4GAKA1zh4AAKMl6LQRUpK5LHYiFSN3WxUjhtIBAOJHMAIANIuxYhTHSnKxHzOBilHE5bo5tQEA4sdQOgCAEetCCKmoGAUes/+3pW7F/vvZQyTFMMeIC7wCANqBYAQAMFqGtbVVvUliZSZcMCq6yNxaaXOOEUPpAADtkNKP1crLyzVr1izl5OQoLy9Ps2fPVk1NTUzP9Xq9mjFjhiwWi1577bVUdhMAICnmawol9UKqcQylswbPMYq8XDcVIwBA/FJaMZo1a5YOHTqkZcuWyel06tZbb9Vtt92mxYsXt/ncxx9/XBaLpc12AIAkCRco+lwqXVPmv2/LlHb9PnL7RI8Zw1Lfvgu5NrmbWsKRJKVZfRd4TWZgAwB0NSkLRlu3btXSpUu1du1aTZgwQZK0cOFCzZw5U4888oj69u0b8bmffvqpHn30Ua1bt059+vRJVRcBAEECPozyekxosaZJjp6t2iVzyJovGDW22TJwjpFvOF3gfv8cI4bSAQDil7KP1VavXq28vLyWUCRJ06ZNk9Vq1Zo1ayI+r66uTjfeeKMWLVqkoqKiNo/T2NioqqqqoFtjY9snWABAK4EVoGhD25I5ZM03MsDTFL2dgucYOQMqTKxKBwBIhpSdPUpLS1VYWBi0z263Kz8/X6WlpRGfd/fdd2vKlCm66qqrYjrOggULlJubG3RbsGBBu/oOAF1TQMUoalBJ5pC15tdwx1cxChxKRzACACRD3EPp5s2bp1//+tdR22zdujWhzixZskTLly/Xxo0bY37O/Pnzdc899wTtczgcCR0fALq0wEDhaZKUFb5dUpfrbh72FsdQuiZ3U/ihdP4XbX+/AABdTtzB6N5779Utt9wStc3gwYNVVFSkI0eOBO13uVwqLy+POERu+fLl2r17t/Ly8oL2X3vttfrWt76lFStWhDzH4XAQhAAgGQKDTtQKThIrM80rzcUzlM7pdkYYShfjqnoAAIQRdzAqKChQQUFBm+0mT56siooKrV+/XuPHj5dkgo/H49GkSZPCPmfevHn6wQ9+ELRv9OjReuyxx3TFFVfE21UAQFziHEqXjGBka/5gK46hdIEVI7vVHrCCqW/pb4IRACB+KVuVbuTIkZo+fbrmzJmjp59+Wk6nU3PnztUNN9zQsiLdgQMHNHXqVP31r3/VxIkTVVRUFLaaNGDAAA0aNChVXQUASMGruUUb2pbM6xjZMto+XrNwc4yChtFZqBgBABKX0hmqL7zwgkaMGKGpU6dq5syZOvfcc/XMM8+0PO50OrV9+3bV1dWlshsAgFhYA0LGiaoYWX3BqO2hdL7rFQVWjPzXMJIYSgcAaI+UXuA1Pz8/6sVci4uL5W1jyENbjwMAksRXvZFO3Bwj31C6eCpGAXOMgitGzf3hvAEASABrmgIADGvAQjbRKjgt1zhK4lA6r0fyuKM2DTfHKHhFOipGAIDEEYwAAEZgxShqBScFQ+naPGYMc4xaFl+IcnFaAAAiSOlQOgBAJ2KLtWKUgqF0UnMwinDtJMVQMWrpT0DFqMdYafhd/vt5o9vXXwDANxbBCABgWOOcY5TMoXSS5I6+AEObc4zCDaUrONfcAABoA0PpAABGrBWjwOsdtVcCQ+ma3E3Rl+tm8QUAQAIIRgAAw9p6WFsEybxeUGAYc9VGbRo4xyj64gvMMQIAxI9gBAAwghZfiKVilIxgFHBMZ0XUpkEVo2jLdTc/BgBAPAhGAAAjsGIUbY5RMq8XFDiUrqkyatM0m7mYq9Ptrxj59jV3zHyJ4WKxAAC0RjACABhBCyHUR2mYoqF0zujBqM05Ri3BiIoRACB+BCMAgBEUUqqiNEzVULrYglHEOUYtQ+moGAEA4kcwAgAYgcPaogWjZK7+FjSUriL08br9LZsxL9dNxQgAkACCEQDAiHlYW4qG0jUeDX28anvLpi8EeeVVnbMuaJ8kyWrz98vjan/fAABdCsEIAGDYYq0Y+U4dAcHIliHZsvy3WE8vgcdsKA19vGpby2ZgCKptMkt7p1sDg1HANlUjAECc7B3dAQDAScIaQ8XIVaeWilHgULrzXk/wmAHBqD6OYOSsDdkXHIyaJGWGP6YlwgVqA/dzkVgA6HKoGAEADKtdsjQPRwt3TSF3k1S7VykbSheuYlQdOpROCqgYRQpGXipGAID4EIwAAH6+qlHD4dDHanZLXldAZSXJq9LVHwh+zFUXdo6RFKliFHgdJlamAwDEh2AEAPCzRQlG1TuaN3xD6dztP17QqnTHpbqD/vvl60IqP3arGQHuC0ZBF3ilYgQAaAeCEQDAz1fBcdVKzprgx3zVm5brBSUhfARWjCSp4jP/dtknIc19FaKwQ+kCh+VxLSMAQJwIRgAAP2uUOT8t832SeL0ge3bw/aBg9HFI85Zg1ObiC1SMAADxIRgBAPwCKzi1e4MfK1/XvOEbSpeEawXZ0puX9252vDkYueqlw++GNI9aMQpZlQ4AgNgRjAAAfoEVo4pN/m1XnVS5xWxbklgxkqT0Hv7tQ29JTRXSzifNcL7WTaNWjAKH0lExAgDEh+sYAQD8AitGlZv92+XrAxZbSEEw8q1I56ySVl4ZPKQusCkVIwBAihCMAAB+9u7+7cqAilHZKv+2b/GFZK38FlgxkqSjH0ZuGq1iZGOOEQAgcQylAwD4pef5tyu3SJ7mKtH+1/37Lc2fqaViKF1bTZvDT72zPui+pFZD6agYAQDiQ8UIAOAXGFJctdK+l6SM3tKxgKWzfUPWOjAYeZsvLhtxKJ27LildAwB0HQQjAIBfWl7w/Q13SRZb8L4ODEZp1rTI9wODkbO6vb0CAHQxBCMAgF/gUDpJajwa2sYXRlI1xyha08AKUev7gRd4dbW6OC0AAG1I2Ryj8vJyzZo1Szk5OcrLy9Ps2bNVU9P2iWr16tW66KKL1K1bN+Xk5Oi8885TfX19qroJAAgUS0hJesWoZ+xNowUjKkYAgHZIWTCaNWuWtmzZomXLlumNN97QypUrddttt0V9zurVqzV9+nRdcsklKikp0dq1azV37lxZrawRAQAnROuhdOEkOxhl9om5afRgRMUIAJC4lAyl27p1q5YuXaq1a9dqwoQJkqSFCxdq5syZeuSRR9S3b9+wz7v77rt15513at68eS37hg8fnoouAgDCaT2ULpyWoXSu5Bwzs1/MTWOuGLkiVIy83nh6BgDoQlJSilm9erXy8vJaQpEkTZs2TVarVWvWrAn7nCNHjmjNmjUqLCzUlClT1Lt3b51//vn66KOPoh6rsbFRVVVVQbfGxsakfj8A0GV0xFC6zPAfloUTezAKUzFqqmC1OgBARCkJRqWlpSosLAzaZ7fblZ+fr9LS0rDP2bNnjyTpwQcf1Jw5c7R06VKNGzdOU6dO1c6dOyMea8GCBcrNzQ26LViwIHnfDAB0Jen5bbdJyVA6S0xNY158oel46JNrv4q/bwCALiOuYDRv3jxZLJaot23btiXUEY/HI0m6/fbbdeutt2rs2LF67LHHNHz4cP3pT3+K+Lz58+ersrIy6DZ//vyE+gAAXV5mUdttkr0qndUuZRS23U5xVIway0KfXPNlIr0DAHQRcc0xuvfee3XLLbdEbTN48GAVFRXpyJEjQftdLpfKy8tVVBT+pNunj5l8O2rUqKD9I0eO1L59+yIez+FwyOFwRHwcABCHtBzJ3s1c3DWSZFeMJDOcruFwm83aFYxqCUYAgMjiCkYFBQUqKChos93kyZNVUVGh9evXa/z48ZKk5cuXy+PxaNKkSWGfU1xcrL59+2r79u1B+3fs2KEZM2bE000AQHtk9JFqdkV+PCXBqJ90fGObzUIu8GoLvMBrwIdkVIwAAHFKyRyjkSNHavr06ZozZ45KSkq0atUqzZ07VzfccEPLinQHDhzQiBEjVFJSIkmyWCz68Y9/rCeeeEL/+Mc/tGvXLt1///3atm2bZs+enYpuAgDCaWv5bF84SWYwyhkZfr89J+hu7BWjY6GvRcUIABBFSpbrlqQXXnhBc+fO1dSpU2W1WnXttdfqiSeeaHnc6XRq+/btqqvzrxB01113qaGhQXfffbfKy8t15plnatmyZRoyZEiqugkAaK2tVeJ8ASRZc4wkKX9c+P09xgbdjR6MbJLFKnk9ZinxpuPBq+xVbTVfWbIbABBGyoJRfn6+Fi9eHPHx4uJiecOcnObNmxd0HSMAwAkWazDyOCWP2wSS9uoRIRjljw+6GzUYSWY4nbvebFfvknqeZbadNQylAwBElZKhdACATizWoXSS5I6ySEM8up8q2buH7o87GAXcr97h367cJIlKEQAgMoIRACBYt0GRH7PYgsOHM8yFVBNhsUg9xgTvs6ZJ+ROCdgUGIavFKru11cCHwAUYqgIW86n4PDn9BAB8YxGMAADBepwZfr/FLmUPDg5GyaoYSaHD6Qq+JaUFV5ECg1HrFerMzmz/dmDFqHxDMnoIAPgGIxgBAIJlD5Xs2aH7c0ZItozgoXTRrncUr9YLMPS9PKRJYDAKGUYnBa9iF1gxOry8vb0DAHzDEYwAAMEsFinvjND9vhXiAitGyQxGfWf6A5nFLg24LqRJ4HWLgq5h1LIzIBhVbpLqDkhHPox+XSYAAEQwAgCE03q+j+Qf6tZWMEr0+kaOXtLwH5nt4XdLWf1DmrRZMQoMRl63tOWX0qafJ9YfAECXkrLlugEAnVjemNB9vS80X4OCUZjFF+r2S9lRFnCIZtR8qftwaeCNYR+OKxhJ0q4/JNYPAECXQ8UIABCqdcUos69/UYbAOUZNFaHPrfs68ePau0mDvhfx2khxByMAAGJEMAIAhMobHbz0dZ8Z/u3AilFTeehza/elrFsEIwBAqhCMAAChbBnS4H/13w/cDgpGx0Of256KURsIRgCAVCEYAQDCO+0nJiANuU0qmOLfHzSULkzFiGAEAOiEWHwBABBeVn/pyn1Sen7w/raG0nVkMLITjAAAiSEYAQAiyygI3RdtKF3jMclVnbLuUDECAKQKQ+kAAPGxBAylqz8Y/FjV1pQeOi1gGF/gdgtHfug+AABiQDACAMQnsFJT82XwY5WpDUZtVowcYSpcAADEgGAEAIhP4FA6d51UX+q/n+KKUZvBKKMwpccHAHxzEYwAAPGxtgokNXv82x0djNJ7SpbwF4cFACAaghEAID72bsH3a3b7tys+T+mh2wxGFovk6JXSPgAAvplYlQ4AEB979+D7vopR7V6zGEP3U1N26B6ZPfS/N/2vJKlfTr/wjRyFUsPhlPUBAPDNRDACAMTHnh18/9ga8/XrV1J/aKtdFw+5OHqjjAKpMuVdAQB8wzCUDgAQn9bBqHSZWYBh9x87pj+tOViAAQAQPypGAID4WG2SLVNy15v7Xpf0xrDQC7taLKHPbb3P601+/1iZDgCQACpGAID4tZ5n1DoUdaRuAyM/xop1AIAICEYAgPilZbfdpqPkjAy/P7OvZMs4sX0BAHQaBCMAQPxazzM6mUQKRjmjTmw/AACdCsEIABC/1kPpTibdBkq2rND9uQQjAEBkBCMAQPxO5oqRxSLlDA/dTzACAESRsmBUXl6uWbNmKScnR3l5eZo9e7ZqamqiPqe0tFTf+973VFRUpG7dumncuHF65ZXUXxcDABCntJO4YiSFH06XO/rE9wMA0GmkLBjNmjVLW7Zs0bJly/TGG29o5cqVuu2226I+5+abb9b27du1ZMkSbdq0Sddcc42uv/56bdy4MVXdBAAk4mSuGEmh1SFbppQ/oWP6AgDoFFISjLZu3aqlS5fq2Wef1aRJk3Tuuedq4cKFevHFF3Xw4MGIz/v44491xx13aOLEiRo8eLB+9rOfKS8vT+vXr09FNwEAiUrL7egeRNe6YtTzbMmW3jF9AQB0CikJRqtXr1ZeXp4mTPB/Ojdt2jRZrVatWbMm4vOmTJmil156SeXl5fJ4PHrxxRfV0NCgCy64IOJzGhsbVVVVFXRrbGxM5rcDAGjN0bOjexBd64pR4Xkd0w8AQKeRkmBUWlqqwsLgK4/b7Xbl5+ertLQ04vP+/ve/y+l0qmfPnnI4HLr99tv16quvaujQoRGfs2DBAuXm5gbdFixYkLTvBQAQxskejHJGBM8p6ndVx/UFANApxBWM5s2bJ4vFEvW2bdu2hDtz//33q6KiQu+++67WrVune+65R9dff702bdoU8Tnz589XZWVl0G3+/PkJ9wEAEANHr47uQduGzTVfe0+T8sd2bF8AACc9ezyN7733Xt1yyy1R2wwePFhFRUU6cuRI0H6Xy6Xy8nIVFRWFfd7u3bv15JNPavPmzTrttNMkSWeeeaY+/PBDLVq0SE8//XTY5zkcDjkcjni+DQBAe6Wf5BUjSSq+SSpfJ43iwzIAQNviCkYFBQUqKChos93kyZNVUVGh9evXa/z48ZKk5cuXy+PxaNKkSWGfU1dXJ0myWoOLWDabTR6PJ55uAgBSrTNUjOxZ0sRnOroXAIBOIiVzjEaOHKnp06drzpw5Kikp0apVqzR37lzdcMMN6tu3ryTpwIEDGjFihEpKSiRJI0aM0NChQ3X77berpKREu3fv1qOPPqply5bp6quvTkU3AQCJOtnnGAEAEKeUXcfohRde0IgRIzR16lTNnDlT5557rp55xv/JndPp1Pbt21sqRWlpaXrrrbdUUFCgK664QmeccYb++te/6vnnn9fMmTNT1U0AQCI6Q8UIAIA4WLxer7ejOwEA6IT+3k1y14XuLzxfmrpCsljafg1OQQCAk0TKKkYAgG84qkYAgG8QghEAIDEZhW23AQCgkyAYAQASkz0k/P70/BPbDwAAkoBgBABITPdTI+wfdmL7AQBAEhCMAACJyR4afj/BCADQCRGMAACJiVgxirAfAICTGMEIAJAYhtIBAL5BCEYAgMRkFEhpucH70ntImb07pj8AALQDwQgAkLjureYZ5Z/VMf0AAKCdCEYAgMS1XoCh19kd0w8AANqJYAQASFzreUY9J3VMPwAAaCeCEQAgcYFByGKXek3uuL4AANAOBCMAQOL6TJcymhdb6H2hWXwBAIBOiGAEAEic1S4NnGW2Rz/UsX0BAKAdLF6v19vRnQAAdGKNx6TDy6UB1wXvt1jafi6nIADASYJgBABIDYIRAKATYSgdAAAAgC7P3tEdAAB8Q1ENAgB0IlSMAAAAAHR5BCMAAAAAXR7BCAAAAECXRzACAAAA0OURjAAAAAB0eQQjAAAAAF0ewQgAAABAl0cwAgAAANDlEYwAAAAAdHn/DzdsFrjaqpYYAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 1000x250 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import logomaker\n",
    "#\n",
    "the_logo = logomaker.Logo(sal_df)\n",
    "the_logo.style_spines(visible=False)\n",
    "the_logo.ax.set_xticks([])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "3nPPAxB9WhiM"
   },
   "outputs": [],
   "source": [
    "plot_saliency(sal_df, negative=negative, start=0, end=90)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "FVyM9R-vWhiM"
   },
   "outputs": [],
   "source": [
    "plot_saliency(sal_df, negative=negative, start=200, end=181)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "q__Cp-iJWhiM"
   },
   "outputs": [],
   "source": [
    "# run extreme evolution:\n",
    "exclude_elements = elements_to_replace\n",
    "evo_df = extreme_evolve(seq, target='export', decreasing=True, exclude=exclude_elements)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "2KGpoJfUWhiN"
   },
   "outputs": [],
   "source": [
    "evo_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "SF3WbdcTWhiN"
   },
   "outputs": [],
   "source": [
    "evo_df.iloc[0]['seq']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "R9DmWWeaWhiN"
   },
   "outputs": [],
   "source": [
    "# nucleotide interactions:\n",
    "arr = connections_array(seq, target='export') # 'expression'/'export'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "MpKA0PCsWhiN"
   },
   "outputs": [],
   "source": [
    "arr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "Gp0z6_D1WhiO"
   },
   "outputs": [],
   "source": [
    "plot_connections(arr, size=9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "4fvlg6PdWhiO"
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
